File: src/main/java/x/withlithum/neoware/util/results/NeoStatus.java
/*
 * SPDX-FileCopyrightText: 2025-2026 WithLithum & contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package x.withlithum.neoware.util.results;

import lombok.Getter;
import org.jetbrains.annotations.Contract;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
public sealed abstract class NeoStatus permits NeoStatus.Ok, NeoStatus.Error {
    public static final class Ok extends NeoStatus {
        private static final Ok INSTANCE = new Ok();

        private Ok() {
        }

        @Override
        public <V> NeoResult<V> withValue(V value) {
            return NeoResult.ok(value);
        }

        @Override
        public boolean equals(Object obj) {
            return obj instanceof Ok;
        }
    }

    public static final class Error extends NeoStatus {
        @Getter
        private final String message;
        @Getter
        @Nullable
        private final Throwable cause;

        private Error(String message, @Nullable Throwable cause) {
            this.message = message;
            this.cause = cause;
        }

        @Override
        public void unwrap() {
            throw new NeoResultException(message, cause);
        }

        @Override
        public <V> NeoResult<V> withValue(V value) {
            return NeoResult.error(message, cause);
        }
    }

    /**
     * Returns an instance of {@link NeoStatus} indicating a successful result. This method always
     * returns the same instance (and thus, it is safe to do identity checks).
     */
    @Contract(pure = true)
    public static NeoStatus ok() {
        return Ok.INSTANCE;
    }

    /**
     * Returns a new result instance that indicates an error, with the specified error message.
     * @param message The message.
     * @return A new instance of {@link Error}.
     */
    public static NeoStatus error(String message) {
        return new Error(message, null);
    }

    /**
     * Returns a new result instance that indicates an error, with the specified error message and
     * an optional cause.
     * @param message The message.
     * @param cause The throwable that caused the failure.
     * @return A new instance of {@link Error}.
     */
    public static NeoStatus error(String message, @Nullable Throwable cause) {
        return new Error(message, cause);
    }

    /**
     * Asserts this instance is a successful status. If it is not, throws
     * {@link NeoResultException}.
     */
    public void unwrap() {
    }

    /**
     * Creates a new instance of {@link NeoResultException}, possibly containing the specified
     * value, corresponding to the current instance.
     *
     * @param value The value to encapsulate.
     * @param <V>   The type of the value to encapsulate.
     * @return An instance of {@link NeoResult} corresponding to the current instance.
     */
    public abstract <V> NeoResult<V> withValue(V value);
}

File: src/main/kotlin/x/withlithum/neoware/util/results/NeoStatus.kt
/*
 * SPDX-FileCopyrightText: 2025-2026 WithLithum & contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package x.withlithum.neoware.util.results

import io.github.oshai.kotlinlogging.KLogger

/**
 * Reports the result of an operation that has no return value.
 */
sealed interface NeoStatus {
    object Ok : NeoStatus {
        override fun unwrap() {
            // Intentionally do nothing since this is a successful result
        }

        override fun <V> withValue(value: V): NeoResult<V> {
            return NeoResult.Ok(value)
        }
    }

    data class Error(val message: String, val cause: Throwable? = null) : NeoStatus {
        fun logWarn(logger: KLogger) {
            logger.warn(cause) { message }
        }

        override fun unwrap() {
            throw NeoResultException(message, cause)
        }

        override fun <V> withValue(value: V): NeoResult<V> {
            return NeoResult.Error(message, cause)
        }
    }

    /**
     * Asserts this instance is a successful status.
     *
     * @throws [NeoResultException] The current status does not indicate success.
     */
    fun unwrap()

    /**
     * Returns a new corresponding instance of [NeoResult] with the specified value.
     */
    fun <V> withValue(value: V): NeoResult<V>
}